db/version_set.cc

```diff
@@ -345,6 +347,385 @@ class FilePicker {
+              // one. Use compressed data structure, has no attribute seqNo
+              assert(!NewestFirstBySeqNo(
+                  files_[0][fp_ctx.curr_index_in_curr_level],
+                  files_[0][fp_ctx.curr_index_in_curr_level - 1]));
+            }
+          }
+          prev_file_ = f;
+        }
+#endif
+        if (cmp_largest == 0) {
+          // cmp_largest is 0, which means the next key will not be in this
+          // file, so stop looking further. Also don't increment megt_iter_
+          // as we may have to look for this key in the next file if we don't
+          // find it in this one
+          break;
+        } else {
+          if (curr_level_ == 0) {
+            // We need to look through all files in level 0
+            ++fp_ctx.curr_index_in_curr_level;
+          }
+          ++mget_iter_;
+        }
+        if (!file_hit) {
+          curr_file_index = (mget_iter_ != current_level_range_.end())
+                                ? file_picker_iter->curr_index_in_curr_level
+                                : curr_file_level_->num_files;
+        }
+      }
+      if (file_hit) {
+        MultiGetRange::Iterator upper_key = mget_iter_;
+        if (cmp_largest == 0) {
+          // Since cmp_largest is 0, mget_iter_ still points to the last key
+          // that falls in this file, instead of the next one. Increment
+          // upper_key so we can set the range properly for SST MultiGet
+          ++upper_key;
+          ++(file_picker_iter->curr_index_in_curr_level);
+          maybe_repeat_key_ = true;
+        }
+        // Set the range for this file
+        current_file_range_ =
+            MultiGetRange(next_file_range, prev_mget_iter_end_, upper_key);
+        returned_file_level_ = curr_level_;
+        return f;
+      }
+    }
+    // Search ended.
+    return nullptr;
+  }
+
+  // getter for current file level
+  // for GET_HIT_L0, GET_HIT_L1 & GET_HIT_L2_AND_UP counts
+  unsigned int GetHitFileLevel() { return hit_file_level_; }
+
+  // Returns true if the most recent "hit file" (i.e., one returned by
+  // GetNextFile()) is at the last index in its level.
+  bool IsHitFileLastInLevel() { return is_hit_file_last_in_level_; }
+
+  const MultiGetRange& CurrentFileRange() { return current_file_range_; }
+
+ private:
+  unsigned int num_levels_;
+  unsigned int curr_level_;
+  unsigned int returned_file_level_;
+  unsigned int hit_file_level_;
+
+  struct FilePickerContext {
+    int32_t search_left_bound;
+    int32_t search_right_bound;
+    unsigned int curr_index_in_curr_level;
+    unsigned int start_index_in_curr_level;
+
+    FilePickerContext(int32_t left, int32_t right)
+        : search_left_bound(left), search_right_bound(right) {}
+
+    FilePickerContext() = default;
+  };
+  std::array<FilePickerContext, MultiGetContext::MAX_BATCH_SIZE>
+      file_picker_ctx_;
+  MultiGetRange* range_;
+  MultiGetRange::Iterator mget_iter_;
+  MultiGetRange::Iterator prev_mget_iter_end_;
```

Comment:
Is it actually the end of the previous SST file?

Reply:
It can be either the last key in the batch that fell into the previous SST file, or the key after that. This is because we may have to search again for the last key in the next SST file, if successive entries of that key are present in both the files.